### PR TITLE
Enable --auto-approve for Policy checks on Remote Backend (Terraform Cloud)

### PR DESCRIPTION
**Pull request by omarismail**
## Background

**Terraform Cloud**

The current functionality with Terraform Cloud (remote backend) is that when there is a _soft-mandatory (can ovveride)_ policy in place, and the policy fails, the User can manually approve via the UI. 

image

**Terraform CLI**

Currently, when this _soft-mandatory_ policy is in place, and it fails during a run, a user can decide to override via the CLI when they use `terraform apply`

image

Selecting `override` works for both the UI and CLI.

## Problem

When the user calls `terraform apply -auto-approve`, the CLI breaks:

image

The CLI fails, but the UI is still waiting for a manual Override. 

## Fix

The goal of `-auto-approve` is to _Skip interactive approval of plan before applying._.

This PR edits the logic in the `backend/remote` for policy checks to look at the `AutoApprove` flag, and determine if it should require a user input or not.

### CODE REVIEW

File: backend/remote/backend_common.go [hunk 1]
```diff
@@ -415,25 +415,27 @@ func (b *Remote) checkPolicy(stopCtx, cancelCtx context.Context, op *backend.Ope
...
+			}
 
-		err = b.confirm(stopCtx, op, opts, r, "override")
-		if err != nil && err != errRunOverridden {
-			return err
+			err = b.confirm(stopCtx, op, opts, r, "override")
+			if err != nil && err != errRunOverridden {
+				return err
+			}
 		}
```
Comment (chrisarcand) [inline backend/remote/backend_common.go:436]: Looks good; can we move this logic up in to the `PolicySoftFailed` `case` block with the unoveriddable logic though? I think the flow felt a little weird even before these changes, with the logic around soft failures being split like this.

**Review by chrisarcand — CHANGES_REQUESTED** [review]
Nice!

File: backend/remote/backend_common.go (hunk 1 above)
Comment (omarismail) [reply backend/remote/backend_common.go]: Agreed, moving it up

**Review by chrisarcand — CHANGES_REQUESTED** [review]

File: backend/remote/backend_common.go [hunk 2]
```diff
@@ -415,32 +415,34 @@ func (b *Remote) checkPolicy(stopCtx, cancelCtx context.Context, op *backend.Ope
 			return fmt.Errorf(msgPrefix + " hard failed.")
 		case tfe.PolicySoftFailed:
 			if op.Type == backend.OperationTypePlan || op.UIOut == nil || op.UIIn == nil ||
-				op.AutoApprove || !pc.Actions.IsOverridable || !pc.Permissions.CanOverride {
+				!pc.Actions.IsOverridable || !pc.Permissions.CanOverride {
 				return fmt.Errorf(msgPrefix + " soft failed.")
```
[the remark this answers is not in the document]
Comment (omarismail) [reply backend/remote/backend_common.go:419]: 👍 , i'm updating the text to give some actionable output with a run link

**Comment by omarismail** [comment]
Updated text:

image

File: backend/remote/backend_common.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (omarismail) [reply backend/remote/backend_common.go]: Updated text to tell the user what to do, and with a link to runs: https://github.com/hashicorp/terraform/pull/27804#issuecomment-783550250

File: backend/remote/backend_common.go [hunk 3]
```diff
@@ -414,33 +414,37 @@ func (b *Remote) checkPolicy(stopCtx, cancelCtx context.Context, op *backend.Ope
 		case tfe.PolicyHardFailed:
 			return fmt.Errorf(msgPrefix + " hard failed.")
 		case tfe.PolicySoftFailed:
+
+			runOverrideOuputMsg := fmt.Sprintf(
```
Comment (chrisarcand) [inline backend/remote/backend_common.go:418]: s/Ouput/Output/

File: backend/remote/backend_common.go [hunk 4]
```diff
@@ -414,33 +414,37 @@ func (b *Remote) checkPolicy(stopCtx, cancelCtx context.Context, op *backend.Ope
...
+				}
+			} else {
+				opts := &terraform.InputOpts{
+					Id:          "override",
+					Query:       "\nDo you want to override the soft failed policy check?",
+					Description: "Only 'override' will be accepted to override.",
+				}
+				err = b.confirm(stopCtx, op, opts, r, "override")
+				if err != nil && err != errRunOverridden {
+					return fmt.Errorf(fmt.Sprintf("Failed to override: %s.\n%s", err.Error(), runOverrideOuputMsg))
```
Comment (chrisarcand) [inline backend/remote/backend_common.go:439]: Something I overlooked before trying this branch out:

Apparently when confirmation not given, the run is automatically discarded, here:

https://github.com/hashicorp/terraform/blob/e966c492813a1a3cbb20afa214c623a055847199/backend/remote/backend_common.go#L536-L560

With the way this error is handled here, you get a "Failed to override" error along with the extra message that isn't really applicable:

image

Here's what a 'I didn't confirm the override' looks like on `master`:

image

We'll want to match this.

**Review by chrisarcand — CHANGES_REQUESTED** [review]
Ran into some issues on the final walk-through, sadly; this is a rather tricksy set of states so you'll want to double check what things look like as someone 1) Not authorized to override 2) Authorized to override and doesn't and 3) Authorized to override and does.

File: backend/remote/backend_common.go (hunk 4 above)
Comment (omarismail) [reply backend/remote/backend_common.go:439]: Aah, good catch! Okay, will update.

**Comment by omarismail** [comment]
Updated to confirm all possible states regarding policy checks and `terraform apply`

All these have a soft-mandatory policy check for `terraform apply`

### Authorized to Override, Manual Discard from UI without `-auto-approve`

image

### Authorized to Override, Manual Approve from UI without `-auto-approve`

image

## Authorized to Override, Manual approve from CLI without `-auto-approve`

image


## Authorized to Override, Manual discard from CLI without `-auto-approve`

image


## Authorized to Override, from CLI with  `-auto-approve`

`terraform apply --auto-approve`

image


### Not Authorized to Override, with and without `-auto-approve`

image

**Review by chrisarcand — APPROVED** [review]

**Pull request merged**